### PR TITLE
PHP Frontend

### PR DESCRIPTION
**Pull request by hoptond**
Data from the database now displays correctly on the portfolio. Not much code to review, try running the unit test and cycling through the projects in the showcase section.

### CODE REVIEW

File: frontend.php [hunk 1]
```diff
@@ -0,0 +1,164 @@
+<?php
+
+/*
+ * Gets a database object to use for database queries. This database's fetch mode is set to FETCH_ASSOC by default.
+ *
+ * @Return returns the object to use for our given queries.
+ */
+function getDBConnection() {
```
Comment (imogenhallett) [inline frontend.php]: You already have a database connection function in cms_functions.  Consider using just one connection function.

File: frontend.php [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    $db = new PDO('mysql:dbname=CMS;host=127.0.0.1','root');
+    $db->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
+    return $db;
+}
+
+function displayAboutMe(PDO $db) {
```
Comment (imogenhallett) [inline frontend.php:11]: This needs a docblock

File: frontend.php [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    $stmt = $db->prepare('SELECT `name`,`title`,`desc` FROM `about`');
+    $stmt->execute();
+    $result = $stmt->fetch();
+    $name = '<h1>' . $result['name'] . '</h1>';
+    $title = '<h2>' . $result['title'] . '</h2>';
+    $desc = '<p>' . $result['desc'] . '</p>';
+    return $name . $title . $desc;
+}
+
+function displayBadges(PDO $db) {
```
Comment (imogenhallett) [inline frontend.php:28]: This needs a docblock

File: frontend.php [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    $output .= '<div class="showcaseviewer"><img src="' . $result['image'] . '">';
+    $output .= '<form class="showcasenav showcaseprev" method="post"><input type="submit" name="prev_' . $id . '" value="&lt" class="showcasenav showcaseprev"></form>';
+    $output .= '<form class="showcasenav showcasenext" method="post"><input type="submit" name="next_' . $id . '" value="&gt" class="showcasenav showcaseprev"></form>';
+    $output .= '<div class="showcasebottom"><a href="' .  $result['link'] . '" class="showcaseview">View Project</a></div>';
+    return $output;
+}
+
+
+
+function getProjectID(PDO $db, $get) {
```
Comment (imogenhallett) [inline frontend.php]: docblock

File: frontend.php [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        if($command[0] == 'prev') {
+            return clampProjectID($db,$id - 1, FALSE);
+        } else if($command[0] == 'next') {
+            return clampProjectID($db,$id + 1, TRUE);
+        }
+    }
+    return getDefaultProject($db);
+}
+
+function getDefaultProject(PDO $db) {
```
Comment (imogenhallett) [inline frontend.php]: docblock

File: frontend.php [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    $stmt = $db->prepare('SELECT `id` FROM `projects`');
+    $stmt->execute();
+    $value = $stmt->fetch();
+    return (int)$value['id'];
+}
+
+
+function displayContactInfo(PDO $db) {
```
Comment (imogenhallett) [inline frontend.php:129]: docblock

**Review by imogenhallett — COMMENTED** [review]
Done :-)

File: frontend.php (hunk 1 above)
Comment (hoptond) [reply frontend.php]: I wanted to keep the cms functions and the frontend functions seperate. Eventually I just settled for extracting the code out of the function to use once in index.php

fixed in 9efcf2a

File: frontend.php (hunk 2 above)
Comment (hoptond) [reply frontend.php:11]: fixed in 9efcf2a

File: frontend.php (hunk 4 above)
Comment (hoptond) [reply frontend.php]: fixed in 9efcf2a

File: frontend.php (hunk 5 above)
Comment (hoptond) [reply frontend.php]: fixed in 9efcf2a

File: frontend.php (hunk 6 above)
Comment (hoptond) [reply frontend.php:129]: fixed in 9efcf2a

File: frontend.php (hunk 3 above)
Comment (hoptond) [reply frontend.php:28]: fixed in 9efcf2a

File: frontend.php [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+ * @param int The target project ID recieved from the GET string.
+ *
+ * @return int The corrected ID of the project we will be displaying.
+ */
+function clampProjectID(PDO $db, int $id, bool $up) {
+    $stmt = $db->prepare('SELECT `id` FROM `projects`');
+    $stmt->execute();
+    $select = $stmt->fetchAll();
+    $array = [];
+    if(in_array($id, $array)) {
```
Comment (mporam) [inline frontend.php]: PSR2

File: frontend.php [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    }
+    return $output;
+}
+
+/*
+ * Clamps the project ID to an existing project in the database. If the ID is higher than any existing number,
+ * we loop back around to the first project ID. If it is lower, we loop back around to the last project ID. If
+ * in the event the autoincrement screws up and we have missing numbers we will round to the nearest existing number.
+ *
+ * @param int The target project ID recieved from the GET string.
```
Comment (mporam) [inline frontend.php]: missing params

File: frontend.php [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    }
+    return roundValueDown($array, $id);
+}
+
+/*
+ * This function rounds the given value upward to the nearest existing value in the array.
+ *
+ * @param array $array The array to search.
+ *
+ * @param numeric $val The value to find the closest match for.
```
Comment (mporam) [inline frontend.php]: params should be grouped together, no line breaks between them

File: frontend.php [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+/*
+ * Clamps the project ID to an existing project in the database. If the ID is higher than any existing number,
+ * we loop back around to the first project ID. If it is lower, we loop back around to the last project ID. If
+ * in the event the autoincrement screws up and we have missing numbers we will round to the nearest existing number.
+ *
+ * @param int The target project ID recieved from the GET string.
+ *
+ * @return int The corrected ID of the project we will be displaying.
+ */
+function clampProjectID(PDO $db, int $id, bool $up) {
```
Comment (mporam) [inline frontend.php]: you shouldnt need to do any of the ID manipulation at all. if this is required, it shows there is a deeper problem with your code. how could you ever have an ID that is not in the db. and why would you not just display an error if you did.

File: index.php [hunk 11]
```diff
@@ -0,0 +1,70 @@
+<?php
+
+require  'frontend.php';
+
+$db = new PDO('mysql:dbname=CMS;host=127.0.0.1','root');
```
Comment (mporam) [inline index.php]: why not use your dbconnect function??

**Review by mporam — CHANGES_REQUESTED** [review]
alot of PSR2 violations and incorrect DOCBLOCKs

File: frontend.php (hunk 10 above)
Comment (hoptond) [reply frontend.php]: The button on each showcase is named prev_$id and next_$id where $id refers to the $id of the project in the database, so clicking next would move to the next entry in the database from that $id. In the real world you'd make the viewer and text change with JS but I don't know enough JS to make this work in the time I have so made this hacky solution work instead.

My reasoning was that if you add a new project, then delete the project in the middle, moving the next project  from the first project will attempt to load the middle project, but that project does not exist anymore in the db so it won't load anything. A similar thing happens exists if you go back from the first project or forward from the last. This was my attempt to fix that.

File: frontend.php (hunk 9 above)
Comment (hoptond) [reply frontend.php]: fixed in 5d4ede8

File: frontend.php (hunk 8 above)
Comment (hoptond) [reply frontend.php]: fixed in 5d4ede8

File: frontend.php (hunk 7 above)
Comment (hoptond) [reply frontend.php]: fixed in 5d4ede8

File: frontend.php (hunk 10 above)
Comment (hoptond) [reply frontend.php]: attempted rework in 3fbee2b

this probably isn't much better but it's late and my brain hurts

**Comment by hoptond** [comment]
changes made

**Comment by hoptond** [comment]
project week over, closing this PR for the moment

**Pull request closed without merging**